src/libraries/System.Private.CoreLib/src/System/IO/Strategies/AppendFileStreamStrategy.cs

```diff
@@ -0,0 +1,193 @@
+
+        public override bool CanWrite => !_fileHandle.IsClosed;
+
+        public override long Length => RandomAccess.GetFileLength(_fileHandle);
+
+        public override long Position
+        {
+            get => Length;
+            set => Seek(value, SeekOrigin.Begin);
+        }
+
+        internal override bool IsAsync => _fileHandle.IsAsync;
+
+        internal override string Name => _fileHandle.Path ?? SR.IO_UnknownFileName;
+
+        internal override bool IsClosed => _fileHandle.IsClosed;
+
+        internal override SafeFileHandle SafeFileHandle => _fileHandle;
+
+        public override ValueTask DisposeAsync()
+        {
+            if (_fileHandle != null && !_fileHandle.IsClosed)
+            {
+                _fileHandle.ThreadPoolBinding?.Dispose();
+                _fileHandle.Dispose();
+            }
+
+            return ValueTask.CompletedTask;
+        }
+
+        internal override void DisposeInternal(bool disposing) => Dispose(disposing);
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing && _fileHandle != null && !_fileHandle.IsClosed)
+            {
+                _fileHandle.ThreadPoolBinding?.Dispose();
+                _fileHandle.Dispose();
+            }
+        }
+
+        public override void Flush() { }  // no buffering = nothing to flush
+
+        public override Task FlushAsync(CancellationToken cancellationToken) => Task.CompletedTask; // no buffering = nothing to flush
+
+        internal override void Flush(bool flushToDisk)
+        {
+            if (flushToDisk && CanWrite)
+            {
+                FileStreamHelpers.FlushToDisk(_fileHandle);
+            }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            if (origin < SeekOrigin.Begin || origin > SeekOrigin.End)
+            {
+                throw new ArgumentException(SR.Argument_InvalidSeekOrigin, nameof(origin));
+            }
+            else if (_fileHandle.IsClosed)
+            {
+                ThrowHelper.ThrowObjectDisposedException_FileClosed();
+            }
+            else if (!CanSeek)
+            {
+                ThrowHelper.ThrowNotSupportedException_UnseekableStream();
+            }
+
+            long length = Length;
+            long newLength = origin == SeekOrigin.Begin ? offset : length + offset;
+
+            if (newLength < 0)
+            {
+                // keep throwing the same exception we did when seek was causing actual offset change
+                FileStreamHelpers.ThrowInvalidArgument(_fileHandle);
+            }
+
+            // Prevent users from overwriting data in a file that was opened in append mode.
+            if (newLength < length)
+            {
+                throw new IOException(SR.IO_SeekAppendOverwrite);
+            }
+
+            FileStreamHelpers.SetFileLength(_fileHandle, newLength);
```

Comment:
I also don't understand this logic.  This program:
```C#
using System;
using System.IO;
using System.Text;

class Program
{
    static void Main()
    {
        const string Path = @"tmp.txt";
        File.Delete(Path);

        using (var fs = new FileStream(Path, FileMode.Append))
        {
            fs.Write(Encoding.ASCII.GetBytes("hello"));
            fs.Position = 1;
            fs.Write(Encoding.ASCII.GetBytes("abc"));
        }

        Console.WriteLine(File.ReadAllText(Path));
    }
}
```
today prints:
```
habco
```
highlighting that a) you are allowed to seek backwards (just not before the position when the file was opened) and b) it's not losing all data after the seek point.